system-configuration/src/network_configuration.rs

```diff
@@ -0,0 +1,468 @@
+                .iter()
+                .map(|s| CFString::new(&format!("{}", s)))
+                .collect();
+            let value = CFArray::from_CFTypes(&addrs);
+            let dictionary = CFDictionary::from_CFType_pairs(&[(key, value)]);
+
+            let path = CFString::new(&format!("Setup:/Network/Service/{}/DNS", self.id()));
+
+            if !store.set(path, dictionary) {
+                return false;
+            }
+        }
+
+        if dns.setup_domain_name.is_some() {
+            let key = CFString::from_static_string("DomainName");
+            let value = CFString::new(dns.setup_domain_name.unwrap().as_str());
+            let dictionary = CFDictionary::from_CFType_pairs(&[(key, value)]);
+
+            let path = CFString::new(&format!("Setup:/Network/Service/{}/DNS", self.id()));
+
+            if !store.set(path, dictionary) {
+                // FIXME: should rollback ?
+                return false;
+            }
+        }
+
+        return true;
+    }
+
+    /// Returns the network interface associated with this network service.
+    pub fn interface(&self) -> Option<SCNetworkInterface> {
+        let interface_ptr = unsafe { SCNetworkServiceGetInterface(self.0) };
+        if interface_ptr.is_null() {
+            None
+        } else {
+            Some(unsafe { SCNetworkInterface::wrap_under_get_rule(interface_ptr) })
+        }
+    }
+}
+
+impl fmt::Display for SCNetworkService {
+    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
+        write!(f, "{:?}", self)
+    }
+}
+
+impl fmt::Debug for SCNetworkService {
+    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
+        write!(
+            f,
+            "SCNetworkService{{ id: {:?}, name: {:?}, enabled: {}, interface: {:?}, dns: {:?} }}",
+            self.id(),
+            self.name(),
+            self.enabled(),
+            self.interface(),
+            self.dns()
+        )
+    }
+}
+
+
+declare_TCFType!{
+    /// Network interface object.
+    SCNetworkInterface, SCNetworkInterfaceRef
+}
+
+impl_TCFType!(
+    SCNetworkInterface,
+    SCNetworkInterfaceRef,
+    SCNetworkInterfaceGetTypeID
+);
+
+
+impl SCNetworkInterface {
+    /// Returns primary network interface
+    pub fn global(store: &SCDynamicStore) -> Option<Self> {
+        if let Some(ifname) = global_query(store, "PrimaryInterface") {
+            for iface in SCNetworkInterface::list() {
+                if let Some(bsd_name) = iface.bsd_name() {
+                    if bsd_name == ifname {
+                        return Some(iface);
+                    }
+                }
+            }
+        }
+
+        return None;
+    }
+
+    /// Returns all network-capable interfaces on the system.
+    pub fn list() -> Vec<SCNetworkInterface> {
+        let array: CFArray<SCNetworkInterfaceRef> =
+            unsafe { CFArray::wrap_under_get_rule(SCNetworkInterfaceCopyAll()) };
+
+        array
+            .get_all_values()
+            .iter()
```

Comment:
Is it not possible to remove `.get_all_values()` here and call `.iter()` directly? Then your iterator should be correctly typed from the start, so there should be no need to then unsafely map the pointer to the correct type.